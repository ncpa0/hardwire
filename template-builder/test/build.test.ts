import { describe, expect, it } from "bun:test";
import fs from "node:fs/promises";
import path from "node:path";

const file = (fp: string) => Bun.file(path.resolve(import.meta.dir, fp));

describe("build", () => {
  it("correctly builds html files", async () => {
    if (await file("./tmp").exists()) {
      Bun.spawnSync({
        cwd: path.dirname(import.meta.dir),
        cmd: ["rm", "-rf", "./tmp"],
      });
    }

    const res = Bun.spawnSync({
      cwd: path.dirname(import.meta.dir),
      cmd: [
        "bun",
        "./src/index.tsx",
        "build",
        "--src",
        "./test/pages/page.tsx",
        "--outdir",
        "./test/tmp",
        "--staticdir",
        "./test/tmp/static",
        "--staticurl",
        "/static",
      ],
    });

    if (!res.success) {
      throw new Error(res.stderr.toString("utf-8"));
    }

    const home = file("./tmp/home.html");
    const about = file("./tmp/about.html");
    const products = file("./tmp/products.html");
    const homeSub1 = file("./tmp/home/sub1.html");
    const homeSub2 = file("./tmp/home/sub1/sub2.html");
    const product1 = file("./tmp/products/1.html");
    const product2 = file("./tmp/products/2.html");
    const product3 = file("./tmp/products/3.html");

    expect(await home.text()).toMatchSnapshot();
    expect(await about.text()).toMatchSnapshot();
    expect(await products.text()).toMatchSnapshot();
    expect(await homeSub1.text()).toMatchSnapshot();
    expect(await homeSub2.text()).toMatchSnapshot();
    expect(await product1.text()).toMatchSnapshot();
    expect(await product2.text()).toMatchSnapshot();
    expect(await product3.text()).toMatchSnapshot();

    const dyns = await fs.readdir(path.resolve(import.meta.dir, "./tmp/__dyn"));
    dyns.sort();

    for (const dyn of dyns) {
      const dynFile = file(`./tmp/__dyn/${dyn}`);
      expect(await dynFile.text()).toMatchSnapshot();
    }
  });
});
