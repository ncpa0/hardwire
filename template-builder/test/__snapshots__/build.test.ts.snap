// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`build correctly builds html files 1`] = `
"<html>
  <head>
    <title>Test</title>
    <script type="module" src="/static/assets/js/eeb9b913.js"></script>
    <script type="module" src="/static/assets/js/86bba22c.js"></script>
    <link
      rel="stylesheet"
      href="/static/assets/css/3f2a2224.css"
      type="text/css"
      media="screen"
    />
    <link
      rel="stylesheet"
      href="/static/assets/css/fa9b0f43.css"
      type="text/css"
      media="screen"
    />
  </head>
  <body>
    <nav>
      <ul>
        <li>
          <a href="/home" hx-boost="true" hx-target="#root">Home</a>
        </li>
        <li>
          <a href="/about" hx-boost="true" hx-target="#root">Home</a>
        </li>
        <li>
          <a href="/products" hx-boost="true" hx-target="#root">Home</a>
        </li>
        <li>
          <a href="/products/1" hx-boost="true" hx-target="#root">Home</a>
        </li>
      </ul>
    </nav>
    <div id="root">
      <h1>Home</h1>
      <a href="/home/sub1/sub2" hx-boost="true" hx-target="#s1"></a>
      <div id="s1"></div>
    </div>
  </body>
</html>
"
`;

exports[`build correctly builds html files 2`] = `
"<html>
  <head>
    <title>Test</title>
    <script type="module" src="/static/assets/js/eeb9b913.js"></script>
    <script type="module" src="/static/assets/js/86bba22c.js"></script>
    <link
      rel="stylesheet"
      href="/static/assets/css/3f2a2224.css"
      type="text/css"
      media="screen"
    />
    <link
      rel="stylesheet"
      href="/static/assets/css/fa9b0f43.css"
      type="text/css"
      media="screen"
    />
  </head>
  <body>
    <nav>
      <ul>
        <li>
          <a href="/home" hx-boost="true" hx-target="#root">Home</a>
        </li>
        <li>
          <a href="/about" hx-boost="true" hx-target="#root">Home</a>
        </li>
        <li>
          <a href="/products" hx-boost="true" hx-target="#root">Home</a>
        </li>
        <li>
          <a href="/products/1" hx-boost="true" hx-target="#root">Home</a>
        </li>
      </ul>
    </nav>
    <div id="root">
      <h1>About</h1>
    </div>
  </body>
</html>
"
`;

exports[`build correctly builds html files 3`] = `
"<html>
  <head>
    <title>Test</title>
    <script type="module" src="/static/assets/js/eeb9b913.js"></script>
    <script type="module" src="/static/assets/js/86bba22c.js"></script>
    <link
      rel="stylesheet"
      href="/static/assets/css/3f2a2224.css"
      type="text/css"
      media="screen"
    />
    <link
      rel="stylesheet"
      href="/static/assets/css/fa9b0f43.css"
      type="text/css"
      media="screen"
    />
  </head>
  <body>
    <nav>
      <ul>
        <li>
          <a href="/home" hx-boost="true" hx-target="#root">Home</a>
        </li>
        <li>
          <a href="/about" hx-boost="true" hx-target="#root">Home</a>
        </li>
        <li>
          <a href="/products" hx-boost="true" hx-target="#root">Home</a>
        </li>
        <li>
          <a href="/products/1" hx-boost="true" hx-target="#root">Home</a>
        </li>
      </ul>
    </nav>
    <div id="root">
      <h2>Products</h2>
      <a href="/products/1" hx-boost="true" hx-target="#products">Open 1</a>
      <a href="/products/2" hx-boost="true" hx-target="#products">Open 2</a>
      <a href="/products/3" hx-boost="true" hx-target="#products">Open 3</a>
      <div id="products"></div>
    </div>
  </body>
</html>
"
`;

exports[`build correctly builds html files 4`] = `
"<html>
  <head>
    <title>Test</title>
    <script type="module" src="/static/assets/js/eeb9b913.js"></script>
    <script type="module" src="/static/assets/js/86bba22c.js"></script>
    <link
      rel="stylesheet"
      href="/static/assets/css/3f2a2224.css"
      type="text/css"
      media="screen"
    />
    <link
      rel="stylesheet"
      href="/static/assets/css/fa9b0f43.css"
      type="text/css"
      media="screen"
    />
  </head>
  <body>
    <nav>
      <ul>
        <li>
          <a href="/home" hx-boost="true" hx-target="#root">Home</a>
        </li>
        <li>
          <a href="/about" hx-boost="true" hx-target="#root">Home</a>
        </li>
        <li>
          <a href="/products" hx-boost="true" hx-target="#root">Home</a>
        </li>
        <li>
          <a href="/products/1" hx-boost="true" hx-target="#root">Home</a>
        </li>
      </ul>
    </nav>
    <div id="root">
      <h1>Home</h1>
      <a href="/home/sub1/sub2" hx-boost="true" hx-target="#s1"></a>
      <div id="s1">
        <h2>Sub1</h2>
        <a href="/home/sub1/sub2" hx-boost="true" hx-target="#s2"></a>
        <div id="s2"></div>
      </div>
    </div>
  </body>
</html>
"
`;

exports[`build correctly builds html files 5`] = `
"<html>
  <head>
    <title>Test</title>
    <script type="module" src="/static/assets/js/eeb9b913.js"></script>
    <script type="module" src="/static/assets/js/86bba22c.js"></script>
    <link
      rel="stylesheet"
      href="/static/assets/css/3f2a2224.css"
      type="text/css"
      media="screen"
    />
    <link
      rel="stylesheet"
      href="/static/assets/css/fa9b0f43.css"
      type="text/css"
      media="screen"
    />
  </head>
  <body>
    <nav>
      <ul>
        <li>
          <a href="/home" hx-boost="true" hx-target="#root">Home</a>
        </li>
        <li>
          <a href="/about" hx-boost="true" hx-target="#root">Home</a>
        </li>
        <li>
          <a href="/products" hx-boost="true" hx-target="#root">Home</a>
        </li>
        <li>
          <a href="/products/1" hx-boost="true" hx-target="#root">Home</a>
        </li>
      </ul>
    </nav>
    <div id="root">
      <h1>Home</h1>
      <a href="/home/sub1/sub2" hx-boost="true" hx-target="#s1"></a>
      <div id="s1">
        <h2>Sub1</h2>
        <a href="/home/sub1/sub2" hx-boost="true" hx-target="#s2"></a>
        <div id="s2">
          <h3>Sub2</h3>
        </div>
      </div>
    </div>
  </body>
</html>
"
`;

exports[`build correctly builds html files 6`] = `
"<html>
  <head>
    <title>Test</title>
    <script type="module" src="/static/assets/js/eeb9b913.js"></script>
    <script type="module" src="/static/assets/js/86bba22c.js"></script>
    <link
      rel="stylesheet"
      href="/static/assets/css/3f2a2224.css"
      type="text/css"
      media="screen"
    />
    <link
      rel="stylesheet"
      href="/static/assets/css/fa9b0f43.css"
      type="text/css"
      media="screen"
    />
  </head>
  <body>
    <nav>
      <ul>
        <li>
          <a href="/home" hx-boost="true" hx-target="#root">Home</a>
        </li>
        <li>
          <a href="/about" hx-boost="true" hx-target="#root">Home</a>
        </li>
        <li>
          <a href="/products" hx-boost="true" hx-target="#root">Home</a>
        </li>
        <li>
          <a href="/products/1" hx-boost="true" hx-target="#root">Home</a>
        </li>
      </ul>
    </nav>
    <div id="root">
      <h2>Products</h2>
      <a href="/products/1" hx-boost="true" hx-target="#products">Open 1</a>
      <a href="/products/2" hx-boost="true" hx-target="#products">Open 2</a>
      <a href="/products/3" hx-boost="true" hx-target="#products">Open 3</a>
      <div id="products">
        <span> Product 1 </span>
      </div>
    </div>
  </body>
</html>
"
`;

exports[`build correctly builds html files 7`] = `
"<html>
  <head>
    <title>Test</title>
    <script type="module" src="/static/assets/js/eeb9b913.js"></script>
    <script type="module" src="/static/assets/js/86bba22c.js"></script>
    <link
      rel="stylesheet"
      href="/static/assets/css/3f2a2224.css"
      type="text/css"
      media="screen"
    />
    <link
      rel="stylesheet"
      href="/static/assets/css/fa9b0f43.css"
      type="text/css"
      media="screen"
    />
  </head>
  <body>
    <nav>
      <ul>
        <li>
          <a href="/home" hx-boost="true" hx-target="#root">Home</a>
        </li>
        <li>
          <a href="/about" hx-boost="true" hx-target="#root">Home</a>
        </li>
        <li>
          <a href="/products" hx-boost="true" hx-target="#root">Home</a>
        </li>
        <li>
          <a href="/products/1" hx-boost="true" hx-target="#root">Home</a>
        </li>
      </ul>
    </nav>
    <div id="root">
      <h2>Products</h2>
      <a href="/products/1" hx-boost="true" hx-target="#products">Open 1</a>
      <a href="/products/2" hx-boost="true" hx-target="#products">Open 2</a>
      <a href="/products/3" hx-boost="true" hx-target="#products">Open 3</a>
      <div id="products">
        <span> Product 2 </span>
      </div>
    </div>
  </body>
</html>
"
`;

exports[`build correctly builds html files 8`] = `
"<html>
  <head>
    <title>Test</title>
    <script type="module" src="/static/assets/js/eeb9b913.js"></script>
    <script type="module" src="/static/assets/js/86bba22c.js"></script>
    <link
      rel="stylesheet"
      href="/static/assets/css/3f2a2224.css"
      type="text/css"
      media="screen"
    />
    <link
      rel="stylesheet"
      href="/static/assets/css/fa9b0f43.css"
      type="text/css"
      media="screen"
    />
  </head>
  <body>
    <nav>
      <ul>
        <li>
          <a href="/home" hx-boost="true" hx-target="#root">Home</a>
        </li>
        <li>
          <a href="/about" hx-boost="true" hx-target="#root">Home</a>
        </li>
        <li>
          <a href="/products" hx-boost="true" hx-target="#root">Home</a>
        </li>
        <li>
          <a href="/products/1" hx-boost="true" hx-target="#root">Home</a>
        </li>
      </ul>
    </nav>
    <div id="root">
      <h2>Products</h2>
      <a href="/products/1" hx-boost="true" hx-target="#products">Open 1</a>
      <a href="/products/2" hx-boost="true" hx-target="#products">Open 2</a>
      <a href="/products/3" hx-boost="true" hx-target="#products">Open 3</a>
      <div id="products">
        <span> Product 3 </span>
      </div>
    </div>
  </body>
</html>
"
`;

exports[`build correctly builds html files 9`] = `
"  <div>
    <h3>{{.User.username}}</h3>
{{if (and .User.isPremium (or (gt .User.age 18) (le .User.sinceSignedUp 30)))}} {{range .User.premiumLinks}} <a href="{{.url}}">{{.name}}</a> {{end}} {{end}}
  </div>"
`;
